gadopt/transport_solver.py

```diff
@@ -0,0 +1,372 @@
+
+        log("Using SU advection")
+
+        J = Function(TensorFunctionSpace(self.mesh, "DQ", 1), name="Jacobian")
+        J.interpolate(Jacobian(self.mesh))
+        # Calculate grid Peclet number. Note the use of a lower bound for diffusivity if
+        # a pure advection scenario is considered.
+        Pe = absv(dot(u, J)) / 2 / (self.su_diffusivity + 1e-12)
+        beta_Pe = as_vector([1 / tanh(Pe_i + 1e-6) - 1 / (Pe_i + 1e-6) for Pe_i in Pe])
+        nubar = dot(absv(dot(u, J)), beta_Pe) / 2  # Calculate SU artificial diffusion
+
+        self.eq_attrs["su_nubar"] = nubar
+
+    @abc.abstractmethod
+    def set_equation(self):
+        """Sets up the term contributions in the equation."""
+        raise NotImplementedError
+
+    def set_solver_options(self) -> None:
+        """Sets PETSc solver parameters."""
+        if isinstance(solver_preset := self.solver_parameters, dict):
+            return
+
+        if solver_preset is not None:
+            match solver_preset:
+                case "direct":
+                    self.solver_parameters = direct_energy_solver_parameters.copy()
+                case "iterative":
+                    self.solver_parameters = iterative_energy_solver_parameters.copy()
+                case _:
+                    raise ValueError(f"Solver type '{solver_preset}' not implemented.")
+        elif self.mesh.topological_dimension() == 2:
+            self.solver_parameters = direct_energy_solver_parameters.copy()
+        else:
+            self.solver_parameters = iterative_energy_solver_parameters.copy()
+
+        if DEBUG >= log_level:
+            self.solver_parameters["ksp_monitor"] = None
+        elif INFO >= log_level:
+            self.solver_parameters["ksp_converged_reason"] = None
+
+    def setup_solver(self) -> None:
+        """Sets up the timestepper using specified parameters."""
+        self.ts = self.timestepper(
+            self.equation,
+            self.solution,
+            self.delta_t,
+            solution_old=self.solution_old,
+            solver_parameters=self.solver_parameters,
+            strong_bcs=self.strong_bcs,
+        )
+
+        self._solver_ready = True
+
+    def solver_callback(self) -> None:
+        """Optional instructions to execute right after a solve."""
+        pass
+
+    def solve(self, t: Number = 0, update_forcings: Callable | None = None) -> None:
+        """Advances solver in time."""
+        if not self._solver_ready:
+            self.setup_solver()
+
+        self.ts.advance(t, update_forcings)
+
+        self.solver_callback()
+
+
+class GenericTransportSolver(GenericTransportBase):
+    """Advances in time a generic transport equation.
+
+    All combinations of advection, diffusion, sink, and source terms are handled.
+
+    Note: The solution field is updated in place.
+
+    Args:
+      terms:
+        List of equation terms (refer to terms_mapping)
```

Comment:
Given that this is intended for user-facing code, I think the expected terms should be explicitly detailed in the docstring. The other concern is that depending on the terms, there are different arguments expected in `eq_attrs`? This needs to be explicit as well.

Reply:
What do you think of my little table?

Reply:
Thanks! I've just tweaked its formatting so it renders correctly on the API pages, but exactly the thing I had in mind.